### ISSUE TITLE
Plans for merging master-gpu branch to master

### ISSUE THREAD

**Issue by mxz297** [HPCToolkit/hpctoolkit#144]
@jmellorcrummey @mwkrentel @Jokeren @laiwei-rice 

In the next a few weeks, we hope to merge the master-gpu branch to the master branch. This merge will be a beast. Since I am new to the HPCToolkit code base, I will need a place for tracking the problems I have and explanations I got from corresponding people. That's reason for creating this issue. 

This merge will bring 30K lines of addition and 3K lines of deletion to the master branch. The exact files that will be changed can be found in the following link:

https://gist.github.com/mxz297/4fd1982d1a247e985973ce009f24504a

The first step of my plan is for everyone who may have committed code to the mater-gpu branch to take a look at the above link and then write a summary about his code changes. Note that the merge has over 500 commits and github is only willing to show the first 250 commits, so I cannot easily identify who has changed what. So, I will need each person to identify his changes and write descriptions.

The descriptions I am hoping for should include the following items:
* What's the problems are you solving? 
* What are your steps for solving these problems in a conceptual level?
* What are the source files changed corresponding to each conceptual step?
* What are the existing code you changed for solving your problems? Describe why existing code does not work for you and use the master branch the definition for existing code.

I will then carefully read your descriptions and read the corresponding source code. My goal is to have a decent understanding of the code, identify inconsistent code design, and identify hard-to-understand code as soon as possible. 

It may seem time consuming to type descriptions in a github issue rather than explaining things in person. But to deal with this large amount of code, I absolutely need to track the descriptions and reference them multiple times. So, please write descriptions in detail and I will carefully read them and cross check with the code. 

A concurrent step is to test the merge in a black-box way. This merge should be able to profile normal cpu code (results should be similar to master branch), NVIDIA GPU code (similar to master-gpu), and improved support for openmp (similar to master-gpu). And a typical test workflow is to manually see whether the hpcviewer/hpctraceviewer are showing similar results before and after the merge. 

I think we need to first determine several testing benchmark for profiling for the three major capabilities (normal cpu code, NVIDIA GPU code, and OpenMP support). I can then determine a strategy for splitting the task. 

Finally, I will start with making sure the merge will compile on various platforms.

**Comment by Jokeren** [comment]
Maybe I will merge this branch into master-gpu first. It just contains a few additional features for instruction mix analysis and only has few changes to the existing master-gpu code.

https://github.com/HPCToolkit/hpctoolkit/tree/master-gpu-instruction-analyzer

**Comment by mxz297** [comment]
In my opinion, I am against merging any more features into master-gpu. We are dealing with a monster merge, so let's not create a larger monster, even if it is just slightly larger. 

What I recommend is that you rebase the master-gpu-instruction-analyze branch on the master-gpu branch. After we merge master-gpu into master, it will be trivial to merge master-gpu-instruction-analyze into master because only the "instruction mix analysis" part needs review. Meanwhile, the master-gpu-instruction-analyze branch still has the update-to-date features for analyzing GPU code.

**Comment by Jokeren** [comment]
@mxz297

If I push some commits to master-gpu now, does it affect your merge? Or you will only merge the current head?

**Comment by mxz297** [comment]
@Jokeren You can push new commits to master-gpu branch. It will not impact my merge.

**Comment by Jokeren** [comment]
@mwkrentel , Thanks for the review.

> The new files in src/lib/banal/cuda have NO copyright notices, NO
> comments in the files, NO overall description of what these files do
> and how they fit into the rest of the code. This is sloppy. The
> copyright notices can be fixed with a script, but for the rest, ....

I agree that there are sparse comments in these files. 

For important produces and steps, actually I have written some comments, but perhaps they are not comprehensive and precise. For other functions, some are interfaces to fit Dyninst, and some are just data structures for parsing dot files (e.g. DotCFG.hpp), so I thought they are self-explainable.

We can have a discussion on this so that I would know how many comments I should add to the files.

> In prof-lean, why add assembly code in atomics.h and
> atomic-powerpc.h. Why not use __sync_val_compare_and_swap or some
> other builtin instead?

I think these are some stale files which could be removed. John added them four years ago.

**Comment by mwkrentel** [comment]
I've committed fixes to master-gpu for (4) and (5) above.

As for (7), I couldn't find a version of autotools that produced the
generated files in master-gpu.  The diffs were the smallest with the
versions from hpctoolkit-devtools, AC 2.69, AM 1.15.1 and LT 2.4.6, so
I went with that.

I suggest we stay with the version from hpctoolkit-devetools for now,
finish the merge, then move to the latest automake.

Add to-do item (8).  Running `hpcrun -e nvidia-cuda` produces a spew
of debugging messages that should be removed.

```
hpcrun -e nvidia-cuda ./main
enter cupti_set_monitoring
activity 22 enable succeeded
activity 1 enable succeeded
activity 2 enable succeeded
enter cupti_set_monitoring
activity 3 enable succeeded
activity 38 enable succeeded
...
loaded module id 14, cubin size 4456, cubin 0x4fb4470
cubin_id 14 hash /home/krentel/release/hpctoolkit-main-measurements/cubins/0ebaa933e8de4d407ba56a598a22ebe3.cubin
cubin_id 14 -> hpctoolkit_module_id 26
Runtime push externalId 9223372036854775809 (cb_id = 31)
Runtime push externalId 9223372036854775810 (cb_id = 31)
Runtime push externalId 9223372036854775811 (cb_id = 31)
Runtime push externalId 9223372036854775812 (cb_id = 31)
exit cupti_correlation_callback_cuda
...
Memcpy copy CorrelationId 218
Memcpy copy kind 1
Memcpy copy bytes 4000
Kernel execution deviceId 0
Kernel execution CorrelationId 219
External CorrelationId 9223372036854775820
CorrelationId 221
External CorrelationId 9223372036854775821
CorrelationId 222
External CorrelationId 9223372036854775819
CorrelationId 220
```

**Comment by Jokeren** [comment]
I've committed a fix to master-gpu for (8).

**Comment by mwkrentel** [comment]
From issue #148, two things for @jokeren's to-do list to fix before
merging with master.

(1) The Fatbin.cpp file is sloppy in how it uses DYNINST_USE_CUDA.
Part of the code tests for it and part of the code just assumes it's
always availble.

Please fix this so that the code will build and run (without support
for cubin binaries, of course) when DYNINST_USE_CUDA is false.

(2) There is a type signature mismatch error in hpcrun/sample-sources
for the function hpcrun_set_new_metric_info_and_period().
hpcrun/metrics.h declares it as:

```
int hpcrun_set_new_metric_info_and_period(kind_info_t *kind, const char* name,
                                          MetricFlags_ValFmt_t valFmt, size_t period, metric_desc_properties_t prop);
```

with the first argument of `kind_info_t *`.  But there are several
places in sample-sources where this argument is missing.

A quick grep shows:

```
cuda.c:      hpcrun_set_new_metric_info_and_period(strdup(buffer),

generic.c:        metric_id = hpcrun_set_new_metric_info_and_period(name,

papi.c:      hpcrun_set_new_metric_info_and_period(strdup(buffer),
papi.c: hpcrun_set_new_metric_info_and_period("idleness",

tst.c:  int metric_id = hpcrun_set_new_metric_info_and_period("_TST",
tst.c:      hpcrun_set_new_metric_info_and_period("idleness (ms)",
```

And check the other related functions, anything with the new
`kind_info_t` argument.

```
hpcrun_set_new_metric_info_w_fn
hpcrun_set_new_metric_info
```

**Comment by mxz297** [comment]
Closed via #165
